### PR TITLE
[FLINK-33488] Implement restore tests for Deduplicate node

### PR DESCRIPTION
**Pull request by jnh5y**
## What is the purpose of the change

Implement restore tests for Deduplicate node

## Verifying this change

This change added tests and can be verified as follows:

* Added restore tests for Deduplicate node which verifies the generated compiled plan with the saved compiled plan

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/DeduplicationTestPrograms.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+                            "insert into MySink "
+                                    + "select order_id, user, product, order_time \n"
+                                    + "FROM ("
+                                    + "  SELECT *,"
+                                    + "    ROW_NUMBER() OVER (PARTITION BY product ORDER BY event_time ASC) AS row_num\n"
+                                    + "  FROM MyTable)"
+                                    + "WHERE row_num = 1")
+                    .build();
+
+    static final TableTestProgram DEDUPLICATE_DESC =
```
Comment (jnh5y) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/DeduplicationTestPrograms.java:82]: I'm not completely sold on this test.  I added it to see an example with `Deduplicate(keep=[LastRow]...`.

If we want to remove it, I'm fine with that.  If the example needs an addition to make it make more sense, I'm open to suggestions.

Comment (dawidwys) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/DeduplicationTestPrograms.java:82]: I like the idea of adding the test. The behaviour is so different that I think it's worth having a test for it.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/DeduplicationTestPrograms.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+                                            Row.ofKind(
+                                                    RowKind.UPDATE_AFTER, 3L, "bob", "pen", 3000L),
+                                            Row.of(4L, "bob", "apple", 4000L),
+                                            Row.ofKind(
+                                                    RowKind.UPDATE_AFTER,
+                                                    5L,
+                                                    "fish",
+                                                    "apple",
+                                                    5000L))
+                                    .consumedAfterRestore(
```
Comment (dawidwys) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/DeduplicationTestPrograms.java:121]: Why is there no `banana`?

File: flink-table/flink-table-planner/src/test/resources/restore-tests/stream-exec-deduplicate_1/deduplicate-asc/plan/deduplicate-asc.json [hunk 3]
```diff
@@ -161,7 +210,7 @@
       "table.exec.mini-batch.size" : "-1"
     },
     "uniqueKeys" : [ 2 ],
-    "isRowtime" : false,
+    "isRowtime" : true,
```
Comment (dawidwys) [inline flink-table/flink-table-planner/src/test/resources/restore-tests/stream-exec-deduplicate_1/deduplicate-asc/plan/deduplicate-asc.json:213]: Should we also have tests for `isRowtime` `false`? I believe we should cover all properties of the exec node.

**Review by dawidwys — COMMENTED** [review]
Looks really good! Good job on adding the additional test. I suggested adding one more test, but otherwise +1

File: flink-table/flink-table-planner/src/test/resources/restore-tests/stream-exec-deduplicate_1/deduplicate-asc-proctime/plan/deduplicate-asc-proctime.json [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+    "id" : 11,
+    "type" : "stream-exec-deduplicate_1",
+    "configuration" : {
+      "table.exec.deduplicate.insert-update-after-sensitive-enabled" : "true",
+      "table.exec.deduplicate.mini-batch.compact-changes-enabled" : "false",
+      "table.exec.mini-batch.enabled" : "false",
+      "table.exec.mini-batch.size" : "-1"
+    },
+    "uniqueKeys" : [ 2 ],
+    "isRowtime" : false,
```
Comment (jnh5y) [inline flink-table/flink-table-planner/src/test/resources/restore-tests/stream-exec-deduplicate_1/deduplicate-asc-proctime/plan/deduplicate-asc-proctime.json:180]: Here's the new test covering isRowtime being false!

**Review by dawidwys — APPROVED** [review]
LGTM, let's wait for CI

**Pull request closed without merging**